clang/lib/CodeGen/Targets/SystemZ.cpp

```diff
@@ -532,9 +532,371 @@ bool SystemZTargetCodeGenInfo::isVectorTypeBased(const Type *Ty,
+    QualType Found;
+    // Check the fields.
+    const RecordDecl *RD = RT->getDecl();
+    for (const auto *FD : RD->fields()) {
+      if (Found.isNull())
+        Found = getSingleElementType(FD->getType());
+      else if (Found != getSingleElementType(FD->getType()))
+        return Ty;
+    }
+    return Found.isNull() ? Ty : Found;
+  }
+
+  const RecordType *RT = Ty->getAs<RecordType>();
+
+  if (RT && RT->isStructureOrClassType()) {
+    const RecordDecl *RD = RT->getDecl();
+    QualType Found;
+
+    // If this is a C++ record, check the bases first.
+    if (const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(RD))
+      if (CXXRD->hasDefinition())
+        for (const auto &I : CXXRD->bases()) {
+          QualType Base = I.getType();
+
+          // Empty bases don't affect things either way.
+          if (isEmptyRecord(getContext(), Base, true))
+            continue;
+
+          if (!Found.isNull())
+            return Ty;
+          Found = getSingleElementType(Base);
+        }
+
+    // Check the fields.
+    for (const auto *FD : RD->fields()) {
+      QualType FT = FD->getType();
+
+      // Ignore empty fields.
+      if (isEmptyField(getContext(), FD, true))
+        continue;
+
+      if (!Found.isNull())
+        return Ty;
+
+      // Treat single element arrays as the element.
+      while (const ConstantArrayType *AT =
+                 getContext().getAsConstantArrayType(FT)) {
+        if (AT->getZExtSize() != 1)
+          break;
+        FT = AT->getElementType();
+      }
+
+      Found = getSingleElementType(FT);
+    }
+
+    // Unlike isSingleElementStruct(), trailing padding is allowed.
+    if (!Found.isNull())
+      return Found;
+  }
+
+  return Ty;
+}
+
+QualType ZOSXPLinkABIInfo::getFPTypeOfComplexLikeType(QualType Ty) const {
+  if (const RecordType *RT = Ty->getAsStructureType()) {
+    const RecordDecl *RD = RT->getDecl();
+
+    // Check for non-empty base classes.
+    if (const CXXRecordDecl *CXXRD = dyn_cast<CXXRecordDecl>(RD))
+      if (CXXRD->hasDefinition())
+        for (const auto &I : CXXRD->bases()) {
+          QualType Base = I.getType();
+          if (!isEmptyRecord(getContext(), Base, true))
+            return QualType();
+        }
+
+    // Check for exactly two elements with exactly the same floating point type.
+    // A single-element struct containing only a float, double, or long double
+    // counts as a field of that type. If the struct has one field consisting
+    // of a complex type, it does not count. This design may be somewhat
+    // inconsistent but it matches the behavior of the legacy C compiler.
+    int Count = 0;
+    clang::BuiltinType::Kind ElemKind;
+    QualType RetTy;
+    for (const auto *FD : RD->fields()) {
+      if (Count >= 2)
+        return QualType();
+
+      QualType FTSingleTy = FD->getType();
+      if (isAggregateTypeForABI(FTSingleTy)) {
+        QualType Ty = getSingleElementType(FTSingleTy);
```

Comment:
This reverted a change we've made in the old PR:  we should just use the common `isSingleElementStruct` here instead of creating our own `getSingleElementType`.

Reply:
I added the following code to getSingleElementType to better handle unions https://github.com/llvm/llvm-project/pull/101024/files#diff-9c0e32c4159fbc7d7385ce3313987cd34c21d270603a4724de689f86e325221fR635-R650. Is there a way to implement this with isSingleElementStruct?

Reply:
ping @uweigand

Reply:
ping @uweigand

Reply:
Sorry, it looks like I didn't actually save a review I had written a while ago, but was still pending on GitHub ...

This is all new news to me. Looks like you're explicitly trying to support any nested combination of structs and unions - is this actually what the ABI requires? This whole thing is used solely for detecting the complex-equivalent types, right? It's still a bit weird to me that only in this particular case, all those variations are accepted. I think it would be good to explicitly specify what the ABI (or de-facto ABI, I guess) actually does.

Also, you're bringing back the whole `getSingleElementType` routine, which also differs in *other* aspects (not just union support) from the default implementation.   We had added our own routine for Linux only to remain compatible with some (accidental) choices made by GCC - this is not something that should be copied to another ABI unless also required there.  I had been discussing those differences with Fanbo in the other PR, and we had come to the conclusion that those do not actually match the behavior of the legacy z/OS compiler either.

Reply:
Hi Ulrich, thanks for the reply. We will take some time to work on fleshing out the requirements more